### ISSUE TITLE
Anti pattern: imperative package management

### ISSUE THREAD

**Issue by domenkozar** [nix-dot-dev/nix.dev#17]

**Comment by domenkozar** [comment]
https://stop-using-nix-env.privatevoid.net/

**Comment by fricklerhandwerk** [comment]
This is nicely to the point, and I [asked the author for permission](https://github.com/privatevoid-net/privatevoid-infrastructure/commit/90a788ce33324a6bd8b88ecfa4ebe0917e36b457#r80617570) to use that here.

[I still think](https://discourse.nixos.org/t/depreciate-the-use-of-nix-env-to-install-packages/20139/14) we should strive to unemotionally enumerate the trade-offs, and then make a recommendation or even leave that judgement to the reader.

**Comment by fricklerhandwerk** [comment]
@GuilloteauQ @adfaure @mpoquet @jojothan @augu5te To pick up from our [last meeting](https://github.com/NixOS/nix.dev/issues/284#issuecomment-1198130228) - thanks for the [update directing beginners to nix.dev](https://gitlab.inria.fr/nix-tutorial/nix-tutorial/-/commit/a9c513547d79ee195d4d7e7a739eb66be3204ddc).

It would be great if you could incrementally address removing `nix-env` from plain sight. It usually does not do what users expect or want. See the other comments on this issue for context.

**Comment by willbush** [comment]
We have https://nix.dev/tutorials/first-steps/towards-reproducibility-pinning-nixpkgs is that enough to close this out?

Also, these issues are sort of TODOs on the topic of Anti-patterns:

- https://github.com/NixOS/nix.dev/issues/16
- https://github.com/NixOS/nix.dev/issues/32
- https://github.com/NixOS/nix.dev/issues/76

Perhaps they can be combined into one issue or just closed.

**Comment by fricklerhandwerk** [comment]
@willbush no, I think we should have clear instructions explaining why one should not use `nix-env` except for two things:
- enumerating attributes but skipping failed assertions
- setting NixOS generations, because that's currently required to get working boot menu entries
